File: 0033 Search in Rotated Sorted Array/search_in_rotated_sorted_array.cpp
/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
public:
    int bs(vector<int>& nums, int left, int right, int target) {
        while(left<=right) {
            int mid = left + (right-left)/2;
            if(nums[mid] == target) return mid;
            else if (nums[mid]>target) right = mid-1;
            else left = mid+1;
        }
        return -1;
    }
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        int left=0, right = n-1;
        while(left<=right) {
            int mid = left + (right-left)/2;
            if (nums[mid]>nums[n-1]) {
                left = mid+1;
            } else {
                right = mid-1;
            }
        }
        int ans = bs(nums, 0, left-1, target);
        if(ans!=-1) return ans; 
        return bs(nums, left, n-1, target);
    }
};

/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.java
/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
    public int search(int[] nums, int target) {
        int n = nums.length;
        int st = 0;
        int end = n-1;
        int mid = 0;
        while(st<end) {
            mid = (st+end)/2;
            if(nums[mid]>nums[end]) st = mid+1;
            else end = mid;
        }
        int rot = st;
        int actualMid = 0;
        st = 0;
        end = n-1;
        while(st<=end) {
            mid = st + (end-st)/2;
            actualMid = (mid + rot)%n;
            if(nums[actualMid]==target) return actualMid;
            if(nums[actualMid]<target) st = mid+1;
            else end = mid-1;
        }
        return -1;
    }
}

/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/
